### PR TITLE
avutil/hwcontext_qsv: fix D3D11VA hwmap errors

### PR DESCRIPTION
**Pull request by tong1wu**
The mfxHDLPair information should be put into texture_infos when
deriving from qsv context. Now fix it to make the qsv->d3d11va zero copy
work properly.

### CODE REVIEW

File: libavutil/hwcontext_d3d11va.c [hunk 1]
```diff
@@ -344,6 +344,22 @@ static int d3d11va_create_staging_texture(AVHWFramesContext *ctx)
...
+            }
+        }
+        if (i == FF_ARRAY_ELEMS(supported_formats)) {
+            av_log(ctx, AV_LOG_ERROR, "Unsupported pixel format: %s\n",
+                av_get_pix_fmt_name(ctx->sw_format));
+            return AVERROR(EINVAL);
+        }
+    }
+
     D3D11_TEXTURE2D_DESC texDesc = {
```
Comment (xhaihao) [inline libavutil/hwcontext_d3d11va.c]: Please do not mix declarations and statements

File: libavutil/hwcontext_qsv.c [hunk 2]
```diff
@@ -806,9 +806,15 @@ static int qsv_frames_derive_from(AVHWFramesContext *dst_ctx,
...
             AVD3D11VAFramesContext *dst_hwctx = dst_ctx->hwctx;
-            mfxHDLPair *pair = (mfxHDLPair*)src_hwctx->surfaces[i].Data.MemId;
-            dst_hwctx->texture = (ID3D11Texture2D*)pair->first;
+            dst_hwctx->texture_infos = av_calloc(src_hwctx->nb_surfaces,
+                                            sizeof(*dst_hwctx->texture_infos));
+            for (i = 0; i < src_hwctx->nb_surfaces; i++) {
+                mfxHDLPair* pair = (mfxHDLPair*)src_hwctx->surfaces[i].Data.MemId;
+                dst_hwctx->texture_infos[i].texture = (ID3D11Texture2D*)pair->first;
+                dst_hwctx->texture_infos[i].index = (intptr_t)pair->second;
+            }
```
Comment (xhaihao) [inline libavutil/hwcontext_qsv.c:827]: Could you add a sample command in the git log to demonstrate the issue ?

Comment (tong1wu) [reply libavutil/hwcontext_qsv.c:827]: ok, I will update the patchset.

Comment (tong1wu) [reply libavutil/hwcontext_qsv.c:827]: > Could you add a sample command in the git log to demonstrate the issue ?

I have resubmitted them, thx.

File: libavutil/hwcontext_qsv.c [hunk 3]
```diff
@@ -1287,6 +1311,7 @@ static int qsv_frames_derive_to(AVHWFramesContext *dst_ctx,
         return AVERROR(ENOSYS);
     }
 
+    s->child_frames_ref = av_buffer_ref(dst_ctx->internal->source_frames);
```
Comment (wenbinc-Bin) [inline libavutil/hwcontext_qsv.c]: You need a failure check here.

File: libavutil/hwcontext_qsv.c [hunk 4]
```diff
@@ -900,9 +910,16 @@ static int qsv_map_from(AVHWFramesContext *ctx,
         dst->height  = src->height;
 
        if (child_frames_ctx->device_ctx->type == AV_HWDEVICE_TYPE_D3D11VA) {
+#if CONFIG_D3D11VA
+           AVD3D11VAFramesContext* child_frames_hwctx = child_frames_ctx->hwctx;
```
Comment (wenbinc-Bin) [inline libavutil/hwcontext_qsv.c]: This indent is not aligned.

File: libavutil/hwcontext_qsv.c [hunk 5]
```diff
@@ -806,9 +806,19 @@ static int qsv_frames_derive_from(AVHWFramesContext *dst_ctx,
 #if CONFIG_D3D11VA
     case AV_HWDEVICE_TYPE_D3D11VA:
         {
+            dst_ctx->initial_pool_size = src_ctx->initial_pool_size;
             AVD3D11VAFramesContext *dst_hwctx = dst_ctx->hwctx;
-            mfxHDLPair *pair = (mfxHDLPair*)src_hwctx->surfaces[i].Data.MemId;
-            dst_hwctx->texture = (ID3D11Texture2D*)pair->first;
+            dst_hwctx->texture_infos = av_calloc(src_hwctx->nb_surfaces,
+                                            sizeof(*dst_hwctx->texture_infos));
+            for (i = 0; i < src_hwctx->nb_surfaces; i++) {
```
Comment (wenbinc-Bin) [inline libavutil/hwcontext_qsv.c:818]: I think we should move this loop after this line
`dst_hwctx->BindFlags = qsv_get_d3d11va_bind_flags(src_hwctx->frame_type);`

Comment (tong1wu) [reply libavutil/hwcontext_qsv.c:818]: ok I will update

**Comment by wenbinc-Bin** [comment]
LGTM Thanks.

**Comment by tong1wu** [comment]
@xhaihao I have split them into 5 commits. The first 3 commits are maintained as discussed and I added the last 2 commits. Please help to review the last 2 thanks.

**Comment by tong1wu** [comment]
closed. Merged into upstream.

**Pull request closed without merging**